src/cpp/benders/factories/BendersFactory.cpp

```diff
@@ -34,154 +28,237 @@ BENDERSMETHOD DeduceBendersMethod(size_t coupling_map_size, size_t batch_size,
-                                       &criterion_computation);
+Outerloop::OuterLoopInputData BendersMainFactory::ProcessCriterionInput(
+    const CouplingMap& couplingMap) {
+  const auto fpath = std::filesystem::path(options_.INPUTROOT) /
+                     options_.OUTER_LOOP_OPTION_FILE;
+  // if adequacy_criterion.yml is provided read it
+  if (std::filesystem::exists(fpath)) {
+    return Outerloop::OuterLoopInputFromYaml().Read(fpath);
+  }
+  // else compute criterion for all areas!
+  else {
+    return GetInputFromSubProblem(couplingMap);
+  }
+}
+Outerloop::OuterLoopInputData BendersMainFactory::GetInputFromSubProblem(
+    const CouplingMap& couplingMap) {
+  auto first_subproblem_pair = std::find_if_not(
+      couplingMap.begin(), couplingMap.end(),
+      [this](const auto& in) { return in.first == options_.MASTER_NAME; });
+  if (first_subproblem_pair == couplingMap.end()) {
+    std::ostringstream stream;
+    auto log_location = LOGLOCATION;
+    stream << "Could not find any Subproblem in structure file "
+           << options_.STRUCTURE_FILE << std::endl;
+    benders_loggers_.display_message(log_location + stream.str());
+    throw InvalidStructureFile(
+        PrefixMessage(LogUtils::LOGLEVEL::FATAL, "Benders"), stream.str(),
+        log_location);
+  } else {
+    const auto first_subproblem_name = first_subproblem_pair->first;
+    return PatternsFromSupbProblem(first_subproblem_name);
+  }
+}
+
+Outerloop::OuterLoopInputData BendersMainFactory::PatternsFromSupbProblem(
+    const std::string& first_subproblem_name) const {
+  SolverAbstract::Ptr solver = BuildSolver(first_subproblem_name);
+  const auto all_variables_name = solver->get_col_names();
+  std::set<std::string> unique_areas = UniqueAreas(all_variables_name);
+  Outerloop::OuterLoopInputData ret;
+  // TODO check this value
+  ret.SetCriterionCountThreshold(1);
+  for (const auto& area : unique_areas) {
+    Outerloop::OuterLoopSingleInputData singleInputData(
+        Outerloop::PositiveUnsuppliedEnergy, area, 1);
+    ret.AddSingleData(singleInputData);
+  }
+  solver->free();
+  return ret;
+}
+
+std::set<std::string> BendersMainFactory::UniqueAreas(
+    const std::vector<std::string>& all_variables_name) const {
+  std::set<std::string> unique_areas;
+  std::regex area_regex(
+      "area<([^>]+)>");  // Regular expression to match area<...>
 
+  for (const auto& str : all_variables_name) {
+    std::smatch match;
+    if (std::regex_search(str, match, area_regex)) {
+      unique_areas.insert(match[1]);  // Insert the matched area into the set
+    }
+  }
+  return unique_areas;
+}
+SolverAbstract::Ptr BendersMainFactory::BuildSolver(
+    const std::string& first_subproblem_name) const {
+  SolverFactory factory(logger_);
+  auto solver_log_manager = SolverLogManager(LogReportsName());
+  auto solver = factory.create_solver(
+      options_.SOLVER_NAME, SOLVER_TYPE::CONTINUOUS, solver_log_manager);
+  solver->set_threads(1);
```

Comment:
Put `set_threads(1)` after `read_prob_mps`, this will avoid us some bugs later on when reading from binary saved files...